File: test-server/src/main/java/al/polis/appserver/service/StudentService.java
package al.polis.appserver.service;

import al.polis.appserver.dto.CourseStudentAssocDto;
import al.polis.appserver.dto.LongIdDto;
import al.polis.appserver.dto.SimpleStringFilterDto;
import al.polis.appserver.dto.StudentDto;
import org.springframework.data.domain.Slice;

import java.util.List;

public interface StudentService {
    StudentDto upsertStudent(StudentDto student);

    Slice<StudentDto> filterStudents(SimpleStringFilterDto filter);

    void deleteStudent(LongIdDto studentId);

    void associateStudentToCourse(CourseStudentAssocDto assoc);

    void removeStudentFromCourse(CourseStudentAssocDto assoc);

    StudentDto getStudent(LongIdDto studentId);

}

File: polis-frontend/src/app/services/student.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { 
  StudentDto, 
  SimpleStringFilterDto, 
  LongIdDto, 
  RespSingleDto, 
  RespSliceDto,
  CourseStudentAssocDto
} from '../models/dto.types';

@Injectable({
  providedIn: 'root'
})
export class StudentService extends ApiService {

  // Get all students with pagination and filtering
  filterStudents(filter: SimpleStringFilterDto): Observable<RespSliceDto<StudentDto>> {
    return this.post<RespSliceDto<StudentDto>>('/student/filter', filter);
  }

  // Get a single student by ID
  getStudent(request: LongIdDto): Observable<RespSingleDto<StudentDto>> {
    return this.post<RespSingleDto<StudentDto>>('/student/get', request);
  }

  // Create or update a student
  upsertStudent(student: StudentDto): Observable<RespSingleDto<StudentDto>> {
    return this.post<RespSingleDto<StudentDto>>('/student/upsert', student);
  }

  // Delete a student
  deleteStudent(request: LongIdDto): Observable<RespSingleDto<StudentDto>> {
    return this.post<RespSingleDto<StudentDto>>('/student/delete', request);
  }

  // Associate student to course
  associateStudentToCourse(assoc: CourseStudentAssocDto): Observable<RespSingleDto<void>> {
    return this.post<RespSingleDto<void>>('/associateStudentToCourse', assoc);
  }

  // Remove student from course
  removeStudentFromCourse(assoc: CourseStudentAssocDto): Observable<RespSingleDto<void>> {
    return this.post<RespSingleDto<void>>('/removeStudentFromCourse', assoc);
  }
}
